astropy/utils/tests/test_data.py

```diff
@@ -372,3 +376,37 @@ def test_is_url_in_cache():
 
     download_file(TESTURL, cache=True, show_progress=False)
     assert is_url_in_cache(TESTURL)
+
+
+@contextlib.contextmanager
+def TemporaryTempDir():
+    """Temporarily redirect all temporary files to a temporary directory"""
+    tempdir = tempfile.mkdtemp()
+    old_tempdir = tempfile.tempdir
+    tempfile.tempdir = tempdir
+    try:
+        yield tempdir
+    finally:
+        tempfile.tempdir = old_tempdir
+        shutil.rmtree(tempdir)
+
+
+def test_get_readable_fileobj_cleans_up_temporary_files():
```

Comment:
Do you prefer this version?

    def test_get_readable_fileobj_cleans_up_temporary_files(monkeypatch):
        """checks that get_readable_fileobj leaves no temporary files behind"""
        # Create a 'file://' URL pointing to a path on the local filesystem
        local_filename = get_pkg_data_filename(os.path.join('data', 'local.dat'))
        url = 'file://' + pathname2url(local_filename)

        # Save temporary files to a known location
        tempdir = tempfile.mkdtemp()

        try:
            monkeypatch.setattr(tempfile, 'tempdir', tempdir)

            # Call get_readable_fileobj() as a context manager
            with get_readable_fileobj(url) as fileobj:
                pass
            # Get listing of files in temporary directory
            tempdir_listing = os.listdir(tempdir)
        finally:
            shutil.rmtree(tempdir)

        # Assert that the temporary file was empty after get_readable_fileobj()
        # context manager finished running
        assert len(tempdir_listing) == 0